src/dinit-iostream.cc

```diff
@@ -0,0 +1,759 @@
+{
+    bool r = open_nx(path);
+    if (io_failure()) {
+        throw iostream_system_err(io_failure());
+    }
+    return r;
+}
+
+bool ostream::open_nx(const char *path, const int flags) noexcept
+{
+    fd = bp_sys::open(path, O_WRONLY | flags);
+
+    if (fd < 0) {
+        io_error = errno;
+        return false;
+    }
+    return true;
+}
+
+bool ostream::open(const char *path, const int flags)
+{
+    bool r = open_nx(path, flags);
+    if (io_failure()) {
+        throw iostream_system_err(io_failure());
+    }
+    return r;
+}
+
+bool ostream::open_nx(const char *path, const int flags, const mode_t mode) noexcept
+{
+    fd = bp_sys::open(path, O_WRONLY | flags, mode);
+
+    if (fd < 0) {
+        io_error = errno;
+        return false;
+    }
+    return true;
+}
+
+bool ostream::open(const char *path, const int flags, const mode_t mode)
+{
+    bool r = open_nx(path, flags, mode);
+    if (io_failure()) {
+        throw iostream_system_err(io_failure());
+    }
+    return r;
+}
+
+bool ostream::close_nx() noexcept
+{
+    if (!is_open()) {
+        return false; // Closed already?
+    }
+    if (flush_nx() < 0) {
+        // io_error was set by flush_nx() call
+        return false;
+    }
+    if (bp_sys::close(fd) < 0) {
+        io_error = errno;
+        return false; // Failed
+    }
+
+    fd = -1;
+    return true;
+}
+
+bool ostream::close()
+{
+    if (!is_open()) {
+        throw iostream_internal_err("Attempted to close pre-closed/invalid fd.");
+        return false;
+    }
+    flush(); // Will throw exception itself
+    if (bp_sys::close(fd) < 0) {
+        io_error = errno;
+        throw iostream_system_err(io_failure());
+        return false;
+    }
+
+    fd = -1;
+    return true;
+}
+
+int ostream::current_state() noexcept
+{
+    int bits = 0;
+    if (buffer_failbit) bits |= io_states::buffer_failbit;
+    if (io_error) bits |= io_states::io_error;
+    return bits;
+}
+
+bool ostream::good() noexcept
+{
+    return (current_state() == 0);
+}
+
+bool ostream::buffer_failure() noexcept
+{
+    return (current_state() & buffer_failbit);
+}
+
+int ostream::io_failure() noexcept
+{
+    return io_error;
+}
+
+void ostream::clear() noexcept
+{
+    // Reset all failures
+    buffer_failbit = false;
+    io_error = 0;
+}
+
+ssize_t ostream::flush_nx() noexcept
+{
+    if (!good()) {
+        return -1;
+    }
+    ssize_t r = 0;
+    while (buf->get_length() > 0) {
+        char *ptr = buf->get_ptr(0);
+        unsigned len = buf->get_contiguous_length(ptr);
+        r = bp_sys::write(fd, ptr, len);
+
+        if (r < 0) {
+            io_error = errno;
+            return -1;
+        }
+        buf->consume(r);
+    }
+
+    return r;
+}
+
+ssize_t ostream::flush()
+{
+    ssize_t r = flush_nx();
+    if (buffer_failure()) {
+        throw iostream_internal_err("Cannot use stream's buffer: buffer is nullptr!");
+    }
+    if (io_failure()) {
+        throw iostream_system_err(io_failure());
+    }
```

Comment:
These two checks are duplicated in many places, they should be factored out into a function